geode-core/src/main/java/org/apache/geode/management/internal/operation/OperationManager.java

```diff
@@ -14,75 +14,95 @@
+  private final Map<Class<? extends ClusterManagementOperation>, BiFunction> performers;
   private final OperationHistoryManager historyManager;
   private final Executor executor;
+  private final InternalCache cache;
 
-  public OperationManager(OperationHistoryManager historyManager) {
+  public OperationManager(InternalCache cache, OperationHistoryManager historyManager) {
+    this.cache = cache;
     this.historyManager = historyManager;
     this.executor = LoggingExecutors.newThreadOnEachExecute("CMSOpPerformer");
 
     // initialize the list of operation performers
     performers = new ConcurrentHashMap<>();
+    registerOperation(RebalanceOperation.class, RebalanceOperationPerformer::perform);
   }
 
   /**
    * for use by modules/extensions to install custom cluster management operations
    */
   public <A extends ClusterManagementOperation<V>, V extends JsonSerializable> void registerOperation(
-      Class<A> operationClass, Function<A, V> operationPerformer) {
+      Class<A> operationClass, BiFunction<Cache, A, V> operationPerformer) {
     performers.put(operationClass, operationPerformer);
   }
 
   public <A extends ClusterManagementOperation<V>, V extends JsonSerializable> OperationInstance<A, V> submit(
       A op) {
     String opId = UUID.randomUUID().toString();
 
-    Function<A, V> performer = getPerformer(op);
+    BiFunction<Cache, A, V> performer = getPerformer(op);
     if (performer == null) {
       throw new IllegalArgumentException(String.format("Operation type %s is not supported",
           op.getClass().getSimpleName()));
     }
 
     CompletableFuture<V> future =
-        CompletableFuture.supplyAsync(() -> performer.apply(op), executor);
+        CompletableFuture.supplyAsync(() -> performer.apply(cache, op), executor);
 
-    OperationInstance<A, V> inst = new OperationInstance<>(future, opId, op);
+    OperationInstance<A, V> inst = new OperationInstance<>(future, opId, op, new Date());
 
     // save the Future so we can check on it later
     return historyManager.save(inst);
   }
 
   @SuppressWarnings("unchecked")
-  private <A extends ClusterManagementOperation<V>, V extends JsonSerializable> Function<A, V> getPerformer(
+  private <C extends Cache, A extends ClusterManagementOperation<V>, V extends JsonSerializable> BiFunction<C, A, V> getPerformer(
       A op) {
     return performers.get(op.getClass());
   }
 
   /**
    * looks up the future for an async operation by id
    */
-  @SuppressWarnings("unchecked")
   public <V extends JsonSerializable> CompletableFuture<V> getStatus(String opId) {
     return historyManager.getStatus(opId);
   }
 
+  /**
+   * looks up the start time of an operation by id, or null if not found
+   */
+  public Date getOperationStart(String opId) {
+    return historyManager.getOperationStart(opId);
+  }
+
+  /**
+   * looks up the end time of an operation by id, or null if not found or in progress
```

Comment:
I think this comment is wrong about null being returned if "in progress". I think it will return the future in that case.
Also, do the callers of this method want to handle "null" if opId is not found or should it instead throw an exception (same question for getOperationStart())?

Reply:
refactored to return all the instance information in one call so the caller can decide what to do in one place if the operation id is not found